File: functions-runner/hmac.ts
// Closes layer 3 of advisory GHSA-7428-mvpp-rhr7 (C3): authenticated
// gateway → runner traffic.
//
// The runner verifies an HMAC-SHA256 signature attached to /invoke
// requests by the gateway. Without the shared secret, a cluster-internal
// attacker cannot forge a valid request — even if they reach
// `runner:8000/invoke` directly via cluster networking.
//
// This module mirrors `internal/functions/hmac.go` byte-for-byte. Any
// change to the canonical-message format MUST be applied to both sides.

const TIMESTAMP_HEADER = "X-Eurobase-Timestamp";
const SIGNATURE_HEADER = "X-Eurobase-Signature";
const MIN_SECRET_LEN = 32;

export type VerifyResult =
  | { ok: true }
  // Caller can distinguish missing headers (warn-only in soft mode) from
  // bad signature (always reject) or out-of-window timestamp (reject —
  // possible replay).
  | { ok: false; reason: "missing" | "bad_timestamp" | "out_of_window" | "mismatch" | "secret_too_short" };

export interface VerifyOptions {
  // Maximum allowed clock-skew between the request timestamp and the
  // verifier's local time, in seconds. Default 300 (5 minutes).
  maxSkewSeconds?: number;
  // Override clock for tests.
  now?: () => number;
}

export interface Verifier {
  verify(headers: Headers, opts?: VerifyOptions): Promise<VerifyResult>;
}

// newVerifier returns a Verifier that uses Web Crypto's HMAC-SHA256.
// The secret is keyed once at construction; subsequent verifies reuse
// the imported CryptoKey for performance.
export async function newVerifier(secret: string): Promise<Verifier> {
  if (secret.length < MIN_SECRET_LEN) {
    throw new Error(
      `functions runner HMAC secret must be at least ${MIN_SECRET_LEN} bytes`,
    );
  }
  const enc = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    enc.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );

  return {
    async verify(headers: Headers, opts: VerifyOptions = {}): Promise<VerifyResult> {
      const ts = headers.get(TIMESTAMP_HEADER);
      const sig = headers.get(SIGNATURE_HEADER);
      if (!ts || !sig) return { ok: false, reason: "missing" };

      const tsNum = Number.parseInt(ts, 10);
      if (!Number.isFinite(tsNum)) return { ok: false, reason: "bad_timestamp" };

      const maxSkew = opts.maxSkewSeconds ?? 300;
      const now = opts.now ? opts.now() : Math.floor(Date.now() / 1000);
      const delta = Math.abs(now - tsNum);
      if (delta > maxSkew) return { ok: false, reason: "out_of_window" };

      const msg = canonicalMessage(headers, ts);
      const macBytes = new Uint8Array(
        await crypto.subtle.sign("HMAC", key, enc.encode(msg)),
      );
      const expectedHex = bytesToHex(macBytes);

      if (!constantTimeEqualHex(sig, expectedHex)) {
        return { ok: false, reason: "mismatch" };
      }
      return { ok: true };
    },
  };
}

// canonicalMessage MUST match internal/functions/hmac.go byte-for-byte.
//
// Format:
//
//   v=1
//   ts=<unix-seconds>
//   project=<X-Project-ID>
//   schema=<X-Schema-Name>
//   function=<X-Function-ID>
//   user=<X-User-ID, or empty>
//   email=<X-User-Email, or empty>
//   plan=<X-Plan>
//   requestid=<X-Request-ID>
//
// LF newlines. Empty values are emitted as empty strings, NOT omitted —
// stops a forged header that adds an unset field from mutating the
// canonical form to match a different signature.
export function canonicalMessage(headers: Headers, ts: string): string {
  const get = (k: string) => headers.get(k) ?? "";
  return (
    "v=1\n" +
    "ts=" + ts + "\n" +
    "project=" + get("X-Project-ID") + "\n" +
    "schema=" + get("X-Schema-Name") + "\n" +
    "function=" + get("X-Function-ID") + "\n" +
    "user=" + get("X-User-ID") + "\n" +
    "email=" + get("X-User-Email") + "\n" +
    "plan=" + get("X-Plan") + "\n" +
    "requestid=" + get("X-Request-ID")
  );
}

function bytesToHex(b: Uint8Array): string {
  let s = "";
  for (let i = 0; i < b.length; i++) {
    s += b[i].toString(16).padStart(2, "0");
  }
  return s;
}

// constantTimeEqualHex compares two hex strings in fixed time. Length
// is checked first (a length mismatch is not a secret); the rest of the
// comparison runs over min(len) bytes XORed into an accumulator that's
// only branched-on at the end. Avoids string short-circuit optimisations.
export function constantTimeEqualHex(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

File: internal/functions/hmac.go
package functions

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"
)

// Closes layer 3 of advisory GHSA-7428-mvpp-rhr7 (C3): authenticated
// gateway → runner traffic.
//
// Without this layer, anything inside the cluster (compromised pod,
// side-car, future ingress mistake) could call `functions:8000/invoke`
// directly with forged `X-Project-ID` / `X-Schema-Name` / `X-User-ID`
// headers and run arbitrary functions in arbitrary tenants — bypassing
// the gateway's authentication entirely. Layer 1 (per-tenant DB role)
// limits SQL reach but the attacker still gets to execute any function
// in any tenant.
//
// With this layer, the gateway HMAC-signs the identity headers using a
// shared secret only the gateway and runner know. The runner verifies
// before it does anything else. A cluster-internal attacker without the
// secret cannot forge a valid request.
//
// Signature scheme: HMAC-SHA256 over a canonical message containing the
// signed headers + a unix timestamp. Timestamp ±5 minutes is accepted
// to allow for clock skew while preventing replay long after capture.

const (
	// signedHeaderTimestamp is the timestamp header the gateway adds and
	// the runner reads. Unix seconds in decimal.
	signedHeaderTimestamp = "X-Eurobase-Timestamp"
	// signedHeaderSignature is the hex-encoded HMAC-SHA256 of the
	// canonical message.
	signedHeaderSignature = "X-Eurobase-Signature"
	// minSecretLen is the minimum acceptable shared-secret length in
	// bytes. 32 bytes = 256 bits, matching the SHA-256 block-equivalent
	// security boundary.
	minSecretLen = 32
)

// Signer signs gateway → runner requests with HMAC-SHA256.
type Signer struct {
	secret []byte
}

// NewSigner constructs a Signer. Returns an error if the secret is
// shorter than 32 bytes — short secrets reduce the entropy of the
// HMAC's effective key and trivialise brute-forcing.
func NewSigner(secret string) (*Signer, error) {
	if len(secret) < minSecretLen {
		return nil, fmt.Errorf("functions runner HMAC secret must be at least %d bytes", minSecretLen)
	}
	return &Signer{secret: []byte(secret)}, nil
}

// Sign attaches X-Eurobase-Timestamp and X-Eurobase-Signature headers
// to the request. The timestamp is unix seconds at sign time; the
// signature is HMAC-SHA256 over the canonical message built from the
// already-set identity headers + that timestamp.
//
// All identity headers (X-Project-ID, X-Schema-Name, etc.) must be
// set BEFORE calling Sign. Modifying them after invalidates the
// signature.
func (s *Signer) Sign(h http.Header, now time.Time) {
	ts := strconv.FormatInt(now.Unix(), 10)
	h.Set(signedHeaderTimestamp, ts)
	msg := canonicalMessage(h, ts)
	mac := hmac.New(sha256.New, s.secret)
	mac.Write([]byte(msg))
	h.Set(signedHeaderSignature, hex.EncodeToString(mac.Sum(nil)))
}

// VerifyOptions configures verification.
type VerifyOptions struct {
	// MaxClockSkew is the maximum allowed difference between the
	// timestamp in the request and the verifier's local time. Recommended:
	// 5 * time.Minute. Zero defaults to that.
	MaxClockSkew time.Duration
	// Now overrides time.Now (test-only).
	Now func() time.Time
}

// ErrMissingSignature is returned when the request has no signature
// headers at all.
var ErrMissingSignature = errors.New("missing signature headers")

// ErrTimestampOutOfWindow is returned when the request's timestamp is
// outside the allowed clock-skew window (replay-or-very-skewed).
var ErrTimestampOutOfWindow = errors.New("timestamp out of allowed window")

// ErrSignatureMismatch is returned when the HMAC doesn't verify.
var ErrSignatureMismatch = errors.New("signature mismatch")

// Verify checks the request's signature. Implements the same scheme as
// Sign. Errors are sentinel values so callers can distinguish missing
// signature (e.g. soft-mode warn) from bad signature (always reject).
//
// Constant-time comparison via hmac.Equal protects against timing-side-
// channel signature forgery.
func (s *Signer) Verify(h http.Header, opts VerifyOptions) error {
	skew := opts.MaxClockSkew
	if skew == 0 {
		skew = 5 * time.Minute
	}
	now := time.Now
	if opts.Now != nil {
		now = opts.Now
	}

	ts := h.Get(signedHeaderTimestamp)
	sig := h.Get(signedHeaderSignature)
	if ts == "" || sig == "" {
		return ErrMissingSignature
	}

	tsInt, err := strconv.ParseInt(ts, 10, 64)
	if err != nil {
		return fmt.Errorf("%w: invalid timestamp", ErrTimestampOutOfWindow)
	}
	delta := now().Unix() - tsInt
	if delta < 0 {
		delta = -delta
	}
	if time.Duration(delta)*time.Second > skew {
		return ErrTimestampOutOfWindow
	}

	expected := hmac.New(sha256.New, s.secret)
	expected.Write([]byte(canonicalMessage(h, ts)))
	expectedHex := hex.EncodeToString(expected.Sum(nil))

	// hmac.Equal is constant-time over equal-length inputs.
	if !hmac.Equal([]byte(sig), []byte(expectedHex)) {
		return ErrSignatureMismatch
	}
	return nil
}

// canonicalMessage builds the deterministic byte string the HMAC is
// computed over. Field order is fixed — must match the runner side
// (functions-runner/hmac.ts) byte-for-byte.
//
// Format:
//
//	v=1
//	ts=<unix-seconds>
//	project=<X-Project-ID>
//	schema=<X-Schema-Name>
//	function=<X-Function-ID>
//	user=<X-User-ID, or empty>
//	email=<X-User-Email, or empty>
//	plan=<X-Plan>
//	requestid=<X-Request-ID>
//
// Newlines are LF. Empty values are emitted as empty strings, NOT
// omitted, so a forged header that adds an unset value can't mutate
// the canonical form to match a different signature.
func canonicalMessage(h http.Header, ts string) string {
	return "v=1\n" +
		"ts=" + ts + "\n" +
		"project=" + h.Get("X-Project-ID") + "\n" +
		"schema=" + h.Get("X-Schema-Name") + "\n" +
		"function=" + h.Get("X-Function-ID") + "\n" +
		"user=" + h.Get("X-User-ID") + "\n" +
		"email=" + h.Get("X-User-Email") + "\n" +
		"plan=" + h.Get("X-Plan") + "\n" +
		"requestid=" + h.Get("X-Request-ID")
}
